### DISCUSSION TITLE
`nb::reference_internal` with `nb::tensor` as return type

### DISCUSSION THREAD

**Discussion by qnzhou** [wjakob/nanobind#53] [Q&A] [answered]
Hi Wenzel,

I am trying to expose an internal buffer owned by a C++ class as numpy array using the new `nb::tensor` class.  However, I am not able to get `nb::rv_policy::reference_internal` or `nb::keep_alive<0, 1>` working properly.  Here is a minimal example illustrating the problem:

```c++
struct A {
    std::vector<int> buffer; // The buffer managed by C++ but we want to expose it as numpy array in python
};

nb::class_<A>(m, "A")
    .def(nb::init<>())
    .def_property_readonly("buffer", [](A& A){
        size_t shape[] = {A.buffer.size()};
        return nb::tensor<nb::numpy, int, nb::shape<nb::any>>(A.buffer.data(), 1, shape);
    });  // `nb::rv_policy::reference_internal` is applied by default
```

In python, a memory error can be generated with the following:

```python
def create():
    a = A()
    return a.buffer()

buffer = create() 
```
I thought the `reference_internal` return policy should keep the parent object `a` valid as long as there is a reference to the `buffer` object.  However, it seems to be not the case, and address sanitizer shows that `buffer` points to invalid memory.  Change the return policy to `nb::keep_alive<0, 1>` also does not work, it trigger the following error:
https://github.com/wjakob/nanobind/blob/f3044cf44763e105428e4e0cf8f42d951b9cc997/src/nb_type.cpp#L725-L727

I am a bit confused as it is mentioned in the README that "it is no longer necessary that bound types are weak-referenceable".  Did I miss anything?  Thanks!

**Comment by wjakob** [comment, marked as answer]
The `nb::reference_internal` feature is intended for cases when you are returning a nanobind-bound object (i.e. something created using `nb::class_<>`).

The `nb::tensor` provides an `owner` parameter specifically for this purpose. Could you try if this works?

**Reply by qnzhou** [reply]
> The nb::tensor provides an owner parameter specifically for this purpose. Could you try if this works?

Thanks!  I have also tried `owner` argument, but I am not sure how to obtain the python object corresponding to `this` within the lambda getter.  I think I need `this` as the `owner` for it to work.  The examples in the doc seems to be creating a new python capsules from scratch as owner.

**Reply by qnzhou** [reply]
Actually, from the [pybind11 doc](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#casting-back-and-forth), `owner = nb::cast(&a)` seems to work pretty well.  Please let me know if this is what you have in mind.

Thanks a lot for the help!

**Reply by wjakob** [reply]
Yeah, that should do it.

**Reply by brettc** [reply]
Sorry for jumping in, but I'm very interested in the new tensor interface in nanobind (see #42). 

Adding the `owner` parameter here fixes the problem, but the numpy array returned is read-only. I believe these flags are set by numpy's `from_dlpack` in the `tensor_wrap` function. So I'm not sure how to return an array to python that is writeable. Have I missed something obvious here?

**Reply by qnzhou** [reply]
I have the same question in https://github.com/wjakob/nanobind/discussions/46.  I think numpy's `ndarray.from_dlpack()` disabled the `write` bit by design, so there may not be anything we can do on nanobind side.

**Reply by brettc** [reply]
ah, I did not see #46. I'll continue the conversation there.
